src/lib/mdns/minimal/responders/QueryResponder.h

```diff
@@ -0,0 +1,225 @@
+    }
+
+    bool IsValid() const { return mInfo != nullptr; }
+
+private:
+    Internal::QueryResponderInfo * mInfo;
+};
+
+class QueryResponderIterator : public std::iterator<std::input_iterator_tag, QueryResponderRecord>
+{
+public:
+    QueryResponderIterator() : mAdditionalOnly(false), mCurrent(nullptr), mRemaining(0) {}
+    QueryResponderIterator(bool additionalOnly, Internal::QueryResponderInfo * pos, size_t size) :
+        mAdditionalOnly(additionalOnly), mCurrent(pos), mRemaining(size)
+    {
+        SkipInvalid();
+    }
+    QueryResponderIterator(const QueryResponderIterator & other) = default;
+    QueryResponderIterator & operator=(const QueryResponderIterator & other) = default;
+
+    QueryResponderIterator & operator++()
+    {
+        if (mRemaining != 0)
+        {
+            mCurrent++;
+            mRemaining--;
+        }
+        SkipInvalid();
+        return *this;
+    }
+
+    QueryResponderIterator operator++(int)
+    {
+        QueryResponderIterator tmp(*this);
+        operator++();
+        return tmp;
+    }
+
+    bool operator==(const QueryResponderIterator & rhs) const { return mCurrent == rhs.mCurrent; }
+    bool operator!=(const QueryResponderIterator & rhs) const { return mCurrent != rhs.mCurrent; }
+
+    QueryResponderRecord & operator*() { return *mCurrent; }
+    QueryResponderRecord * operator->() { return mCurrent; }
+
+    Internal::QueryResponderInfo * GetInternal() { return mCurrent; }
+    const Internal::QueryResponderInfo * GetInternal() const { return mCurrent; }
+
+private:
+    bool isCurrentValid() const
+    {
+        if (mCurrent->responder == nullptr)
+        {
+            return false;
+        }
+
+        if (mAdditionalOnly && !mCurrent->reportNowAsAdditional)
+        {
+            return false;
+        }
+
+        return true;
+    }
+
+    /// Skips invalid/not useful values.
+    /// ensures that if mRemaining is 0, mCurrent is nullptr;
+    void SkipInvalid()
+    {
+        while ((mRemaining > 0) && !isCurrentValid())
+        {
+            mRemaining--;
+            mCurrent++;
+        }
+        if (mRemaining == 0)
+        {
+            mCurrent = nullptr;
+        }
+    }
+
+    bool mAdditionalOnly;
+    Internal::QueryResponderInfo * mCurrent;
+    size_t mRemaining;
+};
+
+/// Responds to mDNS queries.
+///
+/// In particular:
+///   - replies data as provided by the underlying responders
+///   - replies to "_services._dns-sd._udp.local."
+///
+/// Maintains a stateful list of 'additional replies' that can be marked/unmarked
+/// for query processing
+class QueryResponderBase : public Responder // "_services._dns-sd._udp.local"
+{
+public:
+    /// Builds a new responder with the given storage for the response infos
+    QueryResponderBase(Internal::QueryResponderInfo * infos, size_t infoSizes);
+    virtual ~QueryResponderBase() {}
+
+    // Setup initial settings (clears all infos and sets up dns-sd query replies)
+    void Init();
```

Comment:
Is this called by contructor, or shall a client also call it prior to other API calls?

Reply:
Only for the constructor. The problem it bypasses is that base class constructor is called before member constructors in child.

So if I would do the init in the actual contructor, I would setup *infos and then all my setup would be cleared by the child constructor calling the Internal::QueryResponderInfo constructors.

That being said, I did not make it 'protected' or such on purpose: it does what it says it does: clears everything and sets up dns-sd service queries. It could be called by a client if we have a need or we can use it in tests.